crates/extensions/c8y_http_proxy/src/tests.rs

```diff
@@ -63,6 +64,256 @@ async fn c8y_http_proxy_requests_the_device_internal_id_on_start() {
+    tokio::spawn(async move {
+        // NOTE: this is done in the background because this call awaits for the response.
+        proxy
+            .upload_log_binary("test.log", "some log content", None)
+            .await
+            .unwrap();
+    });
+
+    // then the upload request received by c8y is related to the internal id
+    c8y.assert_recv(Some(
+        HttpRequestBuilder::post(format!("https://{c8y_host}/event/events/"))
+            .bearer_auth(token)
+            .header("content-type", "application/json")
+            .header("accept", "application/json")
+            .build()
+            .unwrap(),
+    ))
+    .await;
+}
+
+#[tokio::test]
+async fn retry_software_list_once_with_fresh_internal_id() {
+    let c8y_host = "c8y.tenant.io";
+    let device_id = "device-001";
+    let token = "JWT token";
+    let external_id = "external-device-001";
+    let tmp_dir = "/tmp";
+
+    let (mut proxy, mut c8y) =
+        spawn_c8y_http_proxy(c8y_host.into(), device_id.into(), tmp_dir.into(), token).await;
+
+    // Even before any request is sent to the c8y_proxy
+    // the proxy requests over HTTP the internal device id.
+    let init_request = HttpRequestBuilder::get(format!(
+        "https://{c8y_host}/identity/externalIds/c8y_Serial/{device_id}"
+    ))
+    .bearer_auth(token)
+    .build()
+    .unwrap();
+    c8y.assert_recv(Some(init_request)).await;
+
+    // Cumulocity returns the internal device id, after retrying with the fresh jwt token
+    let c8y_response = HttpResponseBuilder::new()
+        .status(200)
+        .json(&InternalIdResponse::new(device_id, external_id))
+        .build()
+        .unwrap();
+    c8y.send(Ok(c8y_response)).await.unwrap();
+
+    // This internal id is then used by the proxy for subsequent requests.
+    // Create  the  software list and publish
+    let c8y_software_list = C8yUpdateSoftwareListResponse::default();
+    tokio::spawn(async move {
+        // NOTE: this is done in the background because this call awaits for the response.
+        proxy.send_software_list_http(c8y_software_list).await
+    });
+
+    let c8y_software_list = C8yUpdateSoftwareListResponse::default();
+    // then the upload request received by c8y is related to the internal id
+    c8y.assert_recv(Some(
+        HttpRequestBuilder::put(format!(
+            "https://{c8y_host}/inventory/managedObjects/{device_id}"
+        ))
+        .header("content-type", "application/json")
+        .header("accept", "application/json")
+        .bearer_auth(token)
+        .json(&c8y_software_list)
+        .build()
+        .unwrap(),
+    ))
+    .await;
+
+    // The software list upload fails because the device identified with internal id not found
+    let c8y_response = HttpResponseBuilder::new()
+        .status(404)
+        .json(&InternalIdResponse::new(device_id, external_id))
+        .build()
+        .unwrap();
+    c8y.send(Ok(c8y_response)).await.unwrap();
+
```

Comment:
Better to assert the request being resent here before sending the next response.

Reply:
done